src/swapd/swap_state.rs

```diff
@@ -1293,35 +1293,92 @@ fn try_bob_cancel_final_to_swap_end(
+                    );
+                    let abort_all = Task::Abort(Abort {
+                        task_target: TaskTarget::AllTasks,
+                        respond: Boolean::False,
+                    });
+                    event.send_sync_service(
+                        runtime.syncer_state.monero_syncer(),
+                        SyncMsg::Task(abort_all.clone()),
+                    )?;
+                    event.send_sync_service(
+                        runtime.syncer_state.bitcoin_syncer(),
+                        SyncMsg::Task(abort_all),
+                    )?;
+                    // remove txs to invalidate outdated states
+                    runtime.txs.remove(&TxLabel::Cancel);
+                    runtime.txs.remove(&TxLabel::Refund);
+                    runtime.txs.remove(&TxLabel::Buy);
+                    runtime.txs.remove(&TxLabel::Punish);
+                    // send swap outcome to farcasterd
+                    Ok(Some(SwapStateMachine::SwapEnd(Outcome::FailureRefund)))
+                }
+                Some(&TxLabel::Punish) => {
+                    let abort_all = Task::Abort(Abort {
+                        task_target: TaskTarget::AllTasks,
+                        respond: Boolean::False,
+                    });
+                    event.send_sync_service(
+                        runtime.syncer_state.monero_syncer(),
+                        SyncMsg::Task(abort_all.clone()),
+                    )?;
+                    event.send_sync_service(
+                        runtime.syncer_state.bitcoin_syncer(),
+                        SyncMsg::Task(abort_all),
+                    )?;
+                    // remove txs to invalidate outdated states
+                    runtime.txs.remove(&TxLabel::Cancel);
+                    runtime.txs.remove(&TxLabel::Refund);
+                    runtime.txs.remove(&TxLabel::Buy);
+                    Ok(Some(SwapStateMachine::SwapEnd(Outcome::FailurePunish)))
+                }
+                _ => Ok(None),
+            }
+        }
+
+        BusMsg::Sync(SyncMsg::Event(SyncEvent::AddressTransaction(AddressTransaction {
+            id,
+            ref hash,
+            incoming,
+            ..
+        }))) if runtime.syncer_state.tasks.watched_addrs.get(&id) == Some(&TxLabel::Cancel)
+            && !incoming =>
+        {
+            let txid = bitcoin::Txid::from_slice(&hash).expect("should serialize to valid txid");
+            if txid
+                != *runtime
+                    .syncer_state
+                    .tasks
+                    .txids
+                    .get(&TxLabel::Refund)
+                    .expect("should be watching Refund on CancelFinal")
+                && Some(txid)
+                    != runtime
+                        .syncer_state
+                        .tasks
+                        .txids
+                        .get(&TxLabel::Punish)
```

Comment:
if the `Punish` txid is already in `syncer_state.tasks`, we already watching this txid (because we've already hit this branch) and don't need to add another task.